app_flutter/lib/service/appengine_cocoon.dart

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+import 'dart:convert';
+
+import 'package:http/http.dart' as http;
+import 'package:fixnum/fixnum.dart';
+
+import 'package:cocoon_service/protos.dart'
+    show Commit, CommitStatus, Stage, Task;
+
+import 'cocoon.dart';
+
+/// CocoonService for interacting with flutter/flutter production build data.
+///
+/// This queries API endpoints that are hosted on AppEngine.
+class AppEngineCocoonService implements CocoonService {
+  /// The Cocoon API endpoint to query
+  ///
+  /// This is the base for all API requests to cocoon
+  static const baseApiUrl = 'https://flutter-dashboard.appspot.com/api';
+
+  http.Client client = http.Client();
+
+  @override
+  Future<List<CommitStatus>> getStats() async {
+    /// This endpoint returns JSON [List<Agent>, List<CommitStatus>]
+    http.Response response = await client.get('$baseApiUrl/public/get-status');
+
+    if (response.statusCode != 200) {
+      throw new HttpException(
+          '$baseApiUrl/public/get-status returned ${response.statusCode}');
+    }
+
+    Map<String, dynamic> jsonResponse = jsonDecode(response.body);
+    assert(jsonResponse != null);
+
+    return _commitStatusesFromJson(jsonResponse['Statuses']);
+  }
+
+  List<CommitStatus> _commitStatusesFromJson(List<dynamic> jsonCommitStatuses) {
```

Comment:
```suggestion
  List<CommitStatus> _commitStatusesFromJson(List<Object> jsonCommitStatuses) {
```

Reply:
Since these are all `Map<String, Object>` would it be better to expand `List<Object>` to be `List<Map<String, Object>>`?

Reply:
You would think so, but this is a place where Dart's reified generics make life a little trickier.

The JSON decoder which instantiates this List/map can't know or enforce the type of all the objects that get place inside of it, so it creates a List<Object> (these are the same as List<dynamic> or List<void> technically). This actually inserts a marker into the object which encodes the values of the generic. So at runtime, this is more like `ListObject` and not `List<Object>`, if you follow me.

This can cause unexpected problems with runtime type checks. A simpler demonstration:

```
List<Object> xs = <Object>[];
xs.add('a'); // OK
xs.add(1); // OK

List<int> ys = <int>[];
ys.add(1); // OK
List<Object> zs = ys; // OK
zs.add('a'); // FAILURE!
```

Reply:
Thank you for the thoughtful explanation! I also read https://dart.dev/guides/language/sound-dart#type-inference which gave some more insight to what's technically going on.

Reply:
I also edited my comment to fix `<>` from being formatted out...